src/gui/tristateaction.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2019  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QWidgetAction>
+
+class QString;
+class QWidget;
+
+class TriStateWidget;
+
+// TriStateWidget is responsible for checkbox state (tri-state) and paint events while
+// TriStateAction will keep in sync with it. This allows connecting with the usual
+// QAction::triggered slot.
+class TriStateAction : public QWidgetAction
+{
+public:
+    TriStateAction(const QString &text, QWidget *parent);
+
+    void setCheckState(Qt::CheckState checkState);
+    void setChecked(Qt::CheckState checkState);  // provided for compatibility
```

Comment:
Sorry, I didn't understand you. You said:
>There are 2 related functions in Qt: `QAction::setChecked(bool)`, `QCheckBox::setCheckState(Qt::CheckState)`. 

But there is `void setChecked(Qt::CheckState checkState);`...

Do I miss something? Doesn't TriStateAction inherit QAction so these two methods should be overridden from the base methods?

Reply:
Yes, but it doesn't make sense to me that a tri-state widget can have a setChecked(bool) setter. The parameter should always be Qt::CheckState.

Reply:
Sorry, I must have confused myself and confused you... and you didn't untangle me in time.

Anyway, all I want is for you to keep the basic interface open and keep it working properly. E.g.:
```c++
QAction *action = new TriStateAction {/*...*/};
action->setChecked(true); // should be handled properly
```

Reply:
That example is what I've trying to prevent in the first version. A tri-state action/widget shouldn't have a setChecked(bool). The parameter should always be Qt::CheckState.

Reply:
I probably should remove the QWidgetAction inheritance and turn it into a private member and also create our own triggered signals.

Reply:
>I've trying to prevent in the first version. 

As I said before you cannot prevent it (since it's valid use case to access it via base class interface that you cannot modify). All you can do is make it to work properly.

Reply:
>I probably should remove the QWidgetAction inheritance

You can't do it...

Reply:
>You can't do it...

I see what you mean now.
setChecked(bool) isn't a virtual method so there is no way to override it.

Reply:
I added a workaround.